### **8. Dragon Net**






In [1]:
# The code is available at https://github.com/claudiashi57/dragonnet

In [2]:
# import related packages
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt;
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.linear_model import LinearRegression

from causaldm.learners.Causal_Effect_Learning.Single_Stage.Dragonnet import *

### MovieLens Data

In [3]:
# Get data
MovieLens_CEL = pd.read_csv("/Users/alinaxu/Documents/CDM/CausalDM/causaldm/data/MovieLens_CEL.csv")
MovieLens_CEL.pop(MovieLens_CEL.columns[0])
MovieLens_CEL

user_id  movie_id  rating   age  Drama  gender_M  \
0         48.0    1193.0     4.0  25.0    1.0       1.0   
1         48.0     919.0     4.0  25.0    1.0       1.0   
2         48.0     527.0     5.0  25.0    1.0       1.0   
3         48.0    1721.0     4.0  25.0    1.0       1.0   
4         48.0     150.0     4.0  25.0    1.0       1.0   
...        ...       ...     ...   ...    ...       ...   
65637   5878.0    3300.0     2.0  25.0    0.0       0.0   
65638   5878.0    1391.0     1.0  25.0    0.0       0.0   
65639   5878.0     185.0     4.0  25.0    0.0       0.0   
65640   5878.0    2232.0     1.0  25.0    0.0       0.0   
65641   5878.0     426.0     3.0  25.0    0.0       0.0   

       occupation_academic/educator  occupation_college/grad student  \
0                               0.0                              1.0   
1                               0.0                              1.0   
2                               0.0                              1.0   
3                               0.0                              1.0   
4                               0.0                              1.0   
...                             ...                              ...   
65637                           0.0                              0.0   
65638                           0.0                              0.0   
65639                           0.0                              0.0   
65640                           0.0                              0.0   
65641                           0.0                              0.0   

       occupation_executive/managerial  occupation_other  \
0                                  0.0               0.0   
1                                  0.0               0.0   
2                                  0.0               0.0   
3                                  0.0               0.0   
4                                  0.0               0.0   
...                                ...               ...   
65637                              0.0               1.0   
65638                              0.0               1.0   
65639                              0.0               1.0   
65640                              0.0               1.0   
65641                              0.0               1.0   

       occupation_technician/engineer  
0                                 0.0  
1                                 0.0  
2                                 0.0  
3                                 0.0  
4                                 0.0  
...                               ...  
65637                             0.0  
65638                             0.0  
65639                             0.0  
65640                             0.0  
65641                             0.0  

[65642 rows x 11 columns]

In [4]:
n = len(MovieLens_CEL)
userinfo_index = np.array([3,5,6,7,8,9,10])
S = MovieLens_CEL.iloc[:, userinfo_index]
SandA = MovieLens_CEL.iloc[:, np.array([3,4,5,6,7,8,9,10])]

In [5]:
test_outputs, train_output = train_and_predict_dragons(MovieLens_CEL['Drama'].to_numpy().reshape(-1,1),MovieLens_CEL['rating'].to_numpy().reshape(-1,1), S.to_numpy(),
                                                       targeted_regularization=True,
                                                       output_dir='/Users/alinaxu/Downloads/dragonnet-master/result',
                                                       knob_loss=dragonnet_loss_binarycross, ratio=1., dragon='dragonnet',
                                                       val_split=0.2, batch_size=64)

I am here making dragonnet


2023-03-11 02:06:28.678437: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
/Users/alinaxu/opt/anaconda3/lib/python3.9/site-packages/keras/optimizers/optimizer_v2/adam.py:110: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super(Adam, self).__init__(name, **kwargs)


/Users/alinaxu/opt/anaconda3/lib/python3.9/site-packages/keras/optimizers/optimizer_v2/gradient_descent.py:108: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super(SGD, self).__init__(name, **kwargs)


KeyboardInterrupt: 

In [37]:
# the output keys
train_output[0].keys()

dict_keys(['q_t0', 'q_t1', 'g', 't', 'y', 'x', 'index', 'eps'])

In [38]:
HTE_Dragonnet = train_output[0]['q_t1'] - train_output[0]['q_t0']

Let's focus on the estimated HTEs for three randomly chosen users:

In [39]:
print("Dragonnet:  ",HTE_Dragonnet[np.array([0,1000,5000])])

Dragonnet:   [0.32294464 0.43536925 0.33986402]


In [40]:
# Calculate the Aaverage Treatment Effect by Dragonnet
ATE_Dragonnet = np.sum(HTE_Dragonnet)/n
print("Choosing Drama instead of Sci-Fi is expected to improve the rating of all users by",round(ATE_Dragonnet,4), "out of 5 points.")

Choosing Drama instead of Sci-Fi is expected to improve the rating of all users by 0.3225 out of 5 points.


**Conclusion:** Choosing Drama instead of Sci-Fi is expected to improve the rating of all users by 0.3225 out of 5 points.

## References

8. Claudia Shi, David M. Blei, and Victor Veitch. Adapting Neural Networks for the Estimation of
Treatment Effects. 33rd Conference on Neural Information Processing Systems (NeurIPS 2019).